File: PythonSTG/script/Bullet/Bullet.py
from abc import abstractmethod
from enum import Enum


class BulletColor(Enum):
    RED = 1
    PURPLE = 2
    BLUE = 3
    LIGHTBLUE = 4
    LIGHTGREEN = 5
    YELLOW = 6
    ORANGE = 7
    WHITE = 8


class Bullet:
    def __init__(self):
        self.position = [0.0, 0.0, 0.0]
        self.acceleration = 0.0
        self.velocity = 0.3
        self.angle = 0.0
        self.incAngle = 0.0
        self.bulletColor = BulletColor.BLUE
        self.rotateAngle = 0.0
        self.rotateWork = [0, 0, 1]
        self.aimToPlayer = False
        self.reboundCount = 0

        self.releaseParticleColor = [0.0, 0.0, 1.0, 1.0]  # RGBA, default color Blue
        self.releaseParticleDensity = 3  # Generate 3 particles after release
        self.releaseParticleLifeTime = 0.5
        self.releaseParticleVelocity = 0.6
        self.releaseParticleSize = 15

    def setInitCoord(self, coord=[0.0, 0.0, 0.0]):
        self.position = coord

    def setAcceleration(self, a):
        self.acceleration = a

    def setVelocity(self, v):
        self.velocity = v

    def setAngle(self, angle = 0.0):
        if angle >= 360:
            angle -= 360
        elif angle < 0:
            angle += 360
        self.angle = angle

    def setAngleAcceleration(self, increaseAngle=0.0):
        self.incAngle = increaseAngle

    def setAimToPlayer(self, aimToPlayer):
        self.aimToPlayer = aimToPlayer

    @abstractmethod
    def setBulletColor(self, color=BulletColor.BLUE):
        pass

    def setRebound(self, rebound=0):
        self.reboundCount = rebound

    def _setReleaseParticleColor(self, color=[0.0, 0.0, 1.0, 1.0]):
        self.releaseParticleColor = color

    def _setReleaseParticleDensity(self, count=3):
        self.releaseParticleDensity = count

    def _setReleaseParticleLifeTime(self, lifeTime=0.5):
        self.releaseParticleLifeTime = lifeTime

    def _setReleaseParticleVelocity(self, v=0.6):
        self.releaseParticleVelocity = v

    def _setReleaseParticleSize(self, size=15):
        self.releaseParticleSize = size

    def _setRenderSize(self, size=[0.1, 0.1, 0.1]):
        self.scaleSize = size

    def _setCollideSize(self, size):
        self.collideSize = size

    # recall function for cpp init
    def _cpp_getInitCoord(self):
        return tuple(self.position)

    def _cpp_getGenerateInfo(self):
        return self.velocity, self.acceleration, self.angle, self.incAngle, self.reboundCount, self.aimToPlayer

    @abstractmethod
    def _cpp_getInitRenderInfo(self):
        pass

    # return self.bulletType, tuple(self.divideInfo), tuple(self.scaleSize)

    def _cpp_getReleaseParticleInfo(self):
        return self.releaseParticleDensity, self.releaseParticleLifeTime, self.releaseParticleVelocity, self.releaseParticleSize, \
               tuple(self.releaseParticleColor)

File: PythonSTGVisualEditor/Sturcture/Bullet.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Xml.Serialization;
namespace PythonSTGVisualEditor.Sturcture
{
    [XmlRoot("Bullet")]
    public class Bullet : IPythonScript, ICloneable
    {
        public enum BulletColor {
            RED, PURPLE, BLUE, LIGHTBLUE, LIGHTGREEN, YELLOW, ORANGE, WHITE
        };

        private const string bulletInitScript = "{0}={1}()\r\n" +
            "{0}.setInitCoord([{2},{3},0.0])\r\n" +
            "{0}.setRebound({4})\r\n" +
            "{0}.setVelocity({5})\r\n" +
            "{0}.setBulletColor(BulletColor.{6})\r\n" +
            "{0}.setAngle({7})\r\n" +
            "{0}.setAngleAcceleration({8})\r\n" +
            "{0}.setAimToPlayer({9})\r\n";
        [XmlElement("posX")]
        public string posX { get; set; }
        [XmlElement("posY")]
        public string posY { get; set; }
        [XmlElement("velocity")]
        public string velocity { get; set; }
        [XmlElement("Acceleration")]
        public string Acceleration { get; set; }
        [XmlElement("angle")]
        public string angle { get; set; }
        [XmlElement("angleAcceleration")]
        public string angleAcceleration { get; set; }

        [XmlElement("aimToPlayer")]
        public bool aimToPlayer { get; set; }
        [XmlElement("reBound")]
        public string reBound { get; set; }
        [XmlElement("bulletColor")]
        public BulletColor bulletColor { get; set; }

        [XmlElement("bulletVarName")]
        public string bulletVarName { get; set; }
        [XmlElement("bulletClassName")]
        public string bulletClassName { get; set; }


        public string GetInitScript() {
            return string.Format(bulletInitScript, 
                bulletVarName, bulletClassName, posX.ToString(), posY.ToString(),
                reBound.ToString(), velocity.ToString(), bulletColor.ToString(), angle.ToString(), angleAcceleration.ToString(), 
                aimToPlayer ? "True" : "False");
        }

        public object Clone()
        {
            Bullet bullet = new Bullet();

            bullet.bulletVarName = bulletVarName;
            bullet.bulletClassName = bulletClassName;

            bullet.posX = posX;
            bullet.posY = posY;
            bullet.velocity = velocity;
            bullet.Acceleration = Acceleration;
            bullet.aimToPlayer = aimToPlayer;
            bullet.angle = angle;
            bullet.angleAcceleration = angleAcceleration;
            bullet.reBound = reBound;
            bullet.bulletColor = bulletColor;
            return bullet;
        }
    }
}
